chanzuckerberg/single-cell-data-portal
# feat(wmg): filter relationships are precomputed and can be calculated without the selection of primary filters

## Reason for Change

- #2785 
- This PR improves the speed of calculating the valid secondary filters given a particular set of query criteria. The new approach no longer requires the selection of primary filters first to determine the valid secondary filters. As a result, users will be able to select secondary filters _first_ to narrow down the set of available tissues (this will be implemented in a separate PR). 

## Changes

- Update the pipeline to generate a new data artifact containing the relationships between filter values found in the cell counts cube. This graph-like structure contains all the necessary information to calculate valid filter options given a set of query criteria.
- Update `find_dim_option_values` to traverse the filter relationships graph and find the intersection of valid filters given all the query criteria. This replaces the prior logic where we repeatedly queried the cell counts cube when excluding each criteria key.
- Update `build_filter_dims_values` to return valid tissue ontology term IDs. This will be used by the frontend to limit the tissues shown in the tissue dropdown (in a later PR).

## Testing steps

- Tested in `rdev`.
- Unit tests were added to test the valid secondary filter calculation logic. Specifically, these unit tests ensure equivalency between the _old_ and _new_ way of calculating these secondary filters.
- Due to the new way of calculating valid secondary filters given a set of query criteria, some tests needed to be temporarily commented out (specifically the secondary filter tests in `test_query.py`). It is not possible to calculate the filter relationships graph given the tight memory constraints in GHA due to the unnecessarily large size of the dynamically generated virtual snapshots. Because we are still unit testing the valid secondary filter calculation using the aforementioned new tests, it is safe to comment these failing tests out for now. Eventually, when we update the virtual test snapshot to be more wieldy, these tests can be added back.

## Notes for Reviewer
